Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp

```diff
@@ -7,34 +7,84 @@
 #include "MantidAPI/AnalysisDataServiceObserver.h"
 #include "MantidPythonInterface/api/AnalysisDataServiceObserverAdapter.h"
 #include "MantidPythonInterface/core/GetPointer.h"
+#include "MantidPythonInterface/core/ReleaseGlobalInterpreterLock.h"
 
 #include <boost/python/bases.hpp>
 #include <boost/python/class.hpp>
+#include <boost/python/make_function.hpp>
+#include <functional>
 
+using namespace std::placeholders;
 using namespace Mantid::API;
 using namespace Mantid::PythonInterface;
 using namespace boost::python;
+using ObserverMethod = void (AnalysisDataServiceObserver::*)(bool);
+
+namespace {
+
+/**
+ * @brief Given an observer method, such as observeAll, callReleasingGIL will ensure the GIL
+ * has been released before calling the observer method. This is to prevent a deadlock that
+ * can occur if an observe method is called at the same time python is being executed for
+ * an ADS observer handle, such as replaceHandle.
+ *
+ * @param self AnalysisDataServiceObserver; the ADS observer object to call the observe method on
+ * @param on bool; whether to turn on or off the observer for the specific method
+ * @param method ObserverMethod; the method to call with the on parameter.
+ */
+void callReleasingGIL(AnalysisDataServiceObserver &self, bool on, ObserverMethod method) {
+  ReleaseGlobalInterpreterLock releaseGil;
+  (self.*method)(on);
+}
+} // namespace
 
 void export_AnalysisDataServiceObserver() {
   boost::python::class_<AnalysisDataServiceObserver, bases<>, AnalysisDataServiceObserverAdapter, boost::noncopyable>(
       "AnalysisDataServiceObserver", "Observes AnalysisDataService notifications: all only")
-      .def("observeAll", &AnalysisDataServiceObserverAdapter::observeAll, (arg("self"), arg("on")),
+      .def("observeAll",
+           make_function(std::bind(callReleasingGIL, _1, _2, &AnalysisDataServiceObserver::observeAll),
+                         default_call_policies(), (arg("self"), arg("on")),
+                         boost::mpl::vector<void, AnalysisDataServiceObserver &, bool>()),
```

Comment:
This is really nice. Can I ask for one more thing to help with readability? We use the same `boost::mpl::vector` everywhere - can we alias it as well so something like

```cpp
using ObserverMethodSignature = boost::mpl::vector<void, AnalysisDataServiceObserver &, bool>
```

above and then you can just do 

```cpp
default_call_policies(), (arg("self"), arg("on")), ObserverMethodSignature()
```

so it's clearer to the caller what that structure is meant for?

Reply:
Yeah that looks much better!